File: exercise/22/solution2.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        if (n == 0) {
            result.push_back("");
        } else {
            for (int i = 0; i < n; i++) {
                vector<string> left = generateParenthesis(i);
                vector<string> right = generateParenthesis(n - i - 1);
                for (int j = 0; j < left.size(); j++) {
                    for (int k = 0; k < right.size(); k++) {
                        result.push_back("(" + left[j] + ")" + right[k]);
                    }
                }
            }
        }
        return result;
    }
};

File: exercise/22/solution2.py
class Solution:
    cache = {0: [""]}
    def generateParenthesis(self, n: int) -> List[str]:
        result = list()
        if n in self.cache:
            return self.cache[n]
        for i in range(n):
            left = self.generateParenthesis(i)
            right = self.generateParenthesis(n - i - 1)
            for j in range(len(left)):
                for k in range(len(right)):
                    result.append('(' + left[j] + ')' + right[k])
        self.cache[n] = result
        return result

